File: scripts/generate_audio_waveform_peaks.mjs
import fs from "fs";
import path from "path";
import { MPEGDecoder } from "mpg123-decoder";

const AUDIO_PATH = "/home/grem3/mahquranapp/public/audio/abdul_basit_murattal/surah_001.mp3";
const WAVEFORM_DIR = "/home/grem3/mahquranapp/public/data/waveforms";
if (!fs.existsSync(WAVEFORM_DIR)) fs.mkdirSync(WAVEFORM_DIR, { recursive: true });
const OUT_PATH = path.join(WAVEFORM_DIR, "abdul_basit_murattal_surah_1.json");

const fileBuffer = fs.readFileSync(AUDIO_PATH);
const decoder = new MPEGDecoder();
await decoder.ready;
const { channelData, sampleRate } = decoder.decode(fileBuffer);
decoder.free();

const samples = channelData[0];
const duration = samples.length / sampleRate;

// 100 peaks per second
const windowSize = Math.floor(sampleRate / 100);
const peaks = [];
const rmsValues = [];

for (let i = 0; i < samples.length; i += windowSize) {
  let maxVal = 0;
  let sumSq = 0;
  let count = 0;
  for (let j = i; j < Math.min(i + windowSize, samples.length); j++) {
    const absVal = Math.abs(samples[j]);
    if (absVal > maxVal) maxVal = absVal;
    sumSq += absVal * absVal;
    count++;
  }
  const rms = count > 0 ? Math.sqrt(sumSq / count) : 0;
  peaks.push(Number(maxVal.toFixed(4)));
  rmsValues.push(Number(rms.toFixed(4)));
}

const waveformData = {
  reciter: "abdul_basit_murattal",
  surah: 1,
  duration: Number(duration.toFixed(3)),
  sampleRate,
  peaksPerSecond: 100,
  peaks,
  rms: rmsValues
};

fs.writeFileSync(OUT_PATH, JSON.stringify(waveformData, null, 2));
console.log(`✓ Successfully extracted ${peaks.length} waveform peaks across ${duration.toFixed(2)}s for Surah 1!`);

File: scripts/generate_audio_waveform_peaks.py
import json
import os
import subprocess
import struct

AUDIO_PATH = "/home/grem3/mahquranapp/public/audio/abdul_basit_murattal/surah_001.mp3"
WAVEFORM_DIR = "/home/grem3/mahquranapp/public/data/waveforms"
os.makedirs(WAVEFORM_DIR, exist_ok=True)
OUT_PATH = os.path.join(WAVEFORM_DIR, "abdul_basit_murattal_surah_1.json")

# Decode MP3 to raw 16kHz mono PCM via ffmpeg
cmd = [
    "ffmpeg", "-y", "-i", AUDIO_PATH,
    "-ac", "1", "-ar", "16000", "-f", "s16le", "-"
]

proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
raw_pcm, _ = proc.communicate()

# Convert raw PCM bytes to 16-bit integers
samples = struct.unpack(f"{len(raw_pcm)//2}h", raw_pcm)
sample_rate = 16000
duration = len(samples) / sample_rate

# Compute 100 peaks per second (160 samples per window = 10ms)
window_size = 160
peaks = []
rms_values = []

for i in range(0, len(samples), window_size):
    chunk = samples[i:i+window_size]
    if not chunk:
        continue
    peak = max(abs(s) for s in chunk) / 32768.0
    rms = (sum(s*s for s in chunk) / len(chunk))**0.5 / 32768.0
    peaks.append(round(peak, 4))
    rms_values.append(round(rms, 4))

waveform_data = {
    "reciter": "abdul_basit_murattal",
    "surah": 1,
    "duration": round(duration, 3),
    "sampleRate": sample_rate,
    "peaksPerSecond": 100,
    "peaks": peaks,
    "rms": rms_values
}

with open(OUT_PATH, "w", encoding="utf-8") as f:
    json.dump(waveform_data, f)

print(f"Generated high-resolution waveform peaks: {len(peaks)} peaks across {duration:.2f}s ({OUT_PATH})")
